plugins/in_tcp/tcp_conn.c

```diff
@@ -50,22 +119,72 @@ static inline int process_pack(struct tcp_conn *conn,
     while (msgpack_unpack_next(&result, pack, size, &off) == MSGPACK_UNPACK_SUCCESS) {
         entry = result.data;
 
+        appended_address_buffer = NULL;
+        source_address = NULL;
+
         ret = flb_log_event_encoder_begin_record(ctx->log_encoder);
 
         if (ret == FLB_EVENT_ENCODER_SUCCESS) {
             ret = flb_log_event_encoder_set_current_timestamp(ctx->log_encoder);
         }
 
+        if (ctx->source_address_key != NULL) {
+            source_address = flb_connection_get_remote_address(conn->connection);
+        }
+
         if (ret == FLB_EVENT_ENCODER_SUCCESS) {
             if (entry.type == MSGPACK_OBJECT_MAP) {
-                ret = flb_log_event_encoder_set_body_from_msgpack_object(
-                        ctx->log_encoder, &entry);
+                if (source_address != NULL) {
+                    msgpack_sbuffer_init(&sbuf);
```

Comment:
Pass the current record offset and size to `append_message_to_record_data`  instead of cloning it in a disposable msgpack buffer.

Reply:
Right. I haven't noticed this could be handled as passing the current record buffer and its size...
Fixed in https://github.com/fluent/fluent-bit/pull/7672/commits/fd36483975b1f7642425b32ab5f9718b21267ca0

Reply:
I haven't noticed that msgpack buffer and previous offset combination can reduce the allocations for expanding msgpack map. I didn't use the current offset because the current offset should be pointed for the next element of a map.